### ISSUE TITLE
Confusing coercion of floats when used as map keys

### ISSUE THREAD

**Issue by anderseknert** [open-policy-agent/opa#4797]
While unlikely to be a problem in real-world policies, this coercion of float-value keys in maps is pretty confusing:
```
% opa eval -fpretty '{1: true, 1.0: false }'
{
  "1": false
}
% opa eval -fpretty '{1.0: false, 1: true }'
{
  "1": true,
  "1.0": false
}
```

**Comment by philipaconrad** [comment]
@anderseknert Is this something you think we can resolve with better docs?

The [JSON spec's section on Objects](https://datatracker.ietf.org/doc/html/rfc7159#section-4) has a grammar definition that requires all keys in a JSON object to be strings, so I assume that's why the OPA implementation coerces floats to string values here-- we're trying to match the JSON spec.

**Comment by srenatus** [comment]
Rego has always been allowing keys to objects that aren't strings, and when the Rego rubber hits the JSON road, we need to do something about that. So we've been converting everything to strings, for better or for worse. There are known clashes, already, when using `"1"` as an object key while also having `1` as key -- we'll need to pick something. But for floats, this just seems particularly confusing. 

But I'm not completely sure that the JSON coercion isn't just one symptom of another underlying problem. Watch this:
```
$ opa eval -fpretty '1.0 == 1'
true
$ opa eval -fpretty 'count({1.0, 1})'             
2
```
It's the same but a set consisting of both has two elements? I don't quite agree. 😅

**Comment by anderseknert** [comment]
Indeed, it isn't the number -> string coercion that's troubling here, but how `1.0 == 1`... maybe, maybe not :)
The example in the first description (which Stephan provided on Slack) was really my "favorite" one, where the _order of idential entries_ in the map determines the number of items in the output 😄

**Comment by anderseknert** [comment]
Javascript does this correctly, I think:
```javascript
JSON.stringify({1.0: 1})
"{\"1\":1}"
JSON.stringify({1.0: 1, 1: 1})
"{\"1\":1}"
JSON.stringify({1: 1, 1.0: 1})
"{\"1\":1}" 
```

**Comment by philipaconrad** [comment]
Perhaps a more important question to dig into: do we want to expose 2x numeric types at Rego-level (floats/ints), or do we want to expose only Javascript-style Number types that work kind of like floats most of the time?

Under the hood, we could still be switching representation dynamically, based on what's most efficient.

**Comment by srenatus** [comment]
> do we want to expose 2x numeric types at Rego-level (floats/ints), or do we want to expose only Javascript-style Number types that work kind of like floats most of the time?

I think at least on the surface, we're trying to uphold the illusion that there's only one number type in Rego.

**Comment by anderseknert** [comment]
Yep, like it or not, I think we'll have to accept that `1 == 1.0` 😄 Let's just make it _consistently_ so. When creating the issue I thought this was isolated to map keys, but as @srenatus have demonstrated, this is obviously not the case. I'll update the title to reflect this.

**Comment by anderseknert** [comment]
The comparison logic was recently unified and improved across OPA, and while it's not impossible that a few edge cases remain, all of the examples provided in this thread now behave as expected, or at least AFAICS. Closing this issue as the general problem described here is solved. Let's create new issues per edge case found, should we run into any.
